src/libraries/System.Net.Http/src/System/Net/Http/DiagnosticsHelper.cs

```diff
@@ -0,0 +1,123 @@
+        {
+            (1, 0) => "1.0",
+            (1, 1) => "1.1",
+            (2, 0) => "2",
+            (3, 0) => "3",
+            _ => httpVersion.ToString()
+        };
+
+        public static bool TryGetErrorType(HttpResponseMessage? response, Exception? exception, out string? errorType)
+        {
+            if (response is not null)
+            {
+                int statusCode = (int)response.StatusCode;
+
+                // In case the status code indicates a client or a server error, return the string representation of the status code.
+                // See the paragraph Status and the definition of 'error.type' in
+                // https://github.com/open-telemetry/semantic-conventions/blob/2bad9afad58fbd6b33cc683d1ad1f006e35e4a5d/docs/http/http-spans.md
+                if (statusCode >= 400 && statusCode <= 599)
+                {
+                    errorType = GetErrorStatusCodeString(statusCode);
+                    return true;
+                }
+            }
+
+            if (exception is null)
+            {
+                errorType = null;
+                return false;
+            }
+
+            Debug.Assert(Enum.GetValues<HttpRequestError>().Length == 12, "We need to extend the mapping in case new values are added to HttpRequestError.");
+            errorType = (exception as HttpRequestException)?.HttpRequestError switch
+            {
+                HttpRequestError.NameResolutionError => "name_resolution_error",
+                HttpRequestError.ConnectionError => "connection_error",
+                HttpRequestError.SecureConnectionError => "secure_connection_error",
+                HttpRequestError.HttpProtocolError => "http_protocol_error",
+                HttpRequestError.ExtendedConnectNotSupported => "extended_connect_not_supported",
+                HttpRequestError.VersionNegotiationError => "version_negotiation_error",
+                HttpRequestError.UserAuthenticationError => "user_authentication_error",
+                HttpRequestError.ProxyTunnelError => "proxy_tunnel_error",
+                HttpRequestError.InvalidResponse => "invalid_response",
+                HttpRequestError.ResponseEnded => "response_ended",
+                HttpRequestError.ConfigurationLimitExceeded => "configuration_limit_exceeded",
+
+                // Fall back to the exception type name in case of HttpRequestError.Unknown or when exception is not an HttpRequestException.
+                _ => exception.GetType().FullName!
+            };
+            return true;
+        }
+
+        private static object[]? s_boxedStatusCodes;
+        private static string[]? s_statusCodeStrings;
+
+#pragma warning disable CA1859 // we explictly box here
+        public static object GetBoxedStatusCode(int statusCode)
+        {
+            object[] boxes = LazyInitializer.EnsureInitialized(ref s_boxedStatusCodes, static () => new object[512]);
```

Comment:
Why 512?

Reply:
The highest standard status code I'm aware of is [511 Network Authentication Required](https://datatracker.ietf.org/doc/html/rfc6585#section-6.1). We can extend the boxed range to 1000 if we think non-standard values are common enough to cache.

Reply:
Shall we reduce the range to 100-399 codes? Shall we do it in this PR? Note that this code has been around for Metrics since 8.0, this PR only moves it to a shared utility type.

Reply:
If this PR is just moving it, don't need to change it here.